rust/tools/authorization-logic/src/souffle/universe_translation.rs

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright 2022 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! This file contains a translation that populates the universe relations.
+//! Universe relations are facts of the form `isSymbol(x)` that are used to
+//! keep datalog rules grounded and finite. This pattern works by putting facts
+//! of the form `isSymbol(x)` on the RHS of a conditional rule where the variable `x`
+//! would otherwise be ungrounded; then the set of symbols is populated by adding unconditional 
+//! facts of the form `isSymbol("specific_symbol").` to populate these universes.
+//!
+//! A problem with using this pattern in the context of authorization logic is that these universes 
+//! need to be shared across principals. Manually using this pattern is also tedius, and makes the 
+//! code less clear. This translation handles both of these problems by:
+//! 1) extending the program to implement the universe relation pattern by: adding universe 
+//!    predicates to the RHS of rules, and adding facts that add elements to universes.
+//! 2) extending the program with facts that use delegation to pass membership of universes across 
+//!    principals.
+//!
+//! This translation is a compiler pass that operates on the data structure described in `ast.rs` 
+//! and produces a new instance of the data structure in `ast.rs`.
+//! It should come before the translation in `lowering_ast_datalog.rs`.
+
+use crate::ast::*;
+use std::collections::HashMap;
+
+// Determine if a function argument or principal name is a constant or 
+// return false if this name is a variable. In the ast, all arguments are
+// represented as strings, and
+// they are constants if and only if they are wrapped in quotes. A better
+// way to represent this in the future might be to add an AST node for
+// arguments with varibales and constants as two possible forms.
+fn is_name_constant(arg: &String) -> bool {
+    return arg.contains("\"")
+}
+
+fn push_option_to_vec<T>(vec: &mut Vec<T>, elt: Option<T>) {
+    if let Some(x) = elt {
+        vec.push(x);
+    }
+}
```

Comment:
Rather than have this take a mutable reference to a `Vec` and possibly mutate it, you could have it take a `Vec` value and return a `Vec`. Because Rust moves by default, it will probably be as efficient as what you have written here, but will fit more in a functional programming model.

Reply:
This is a draft please don't review yet

Reply:
whoops, sorry.

Reply:
I think I'd rather leave this as mutating a borrow since that follows the `push` from the `vec` api if that's ok